python/src/scipp/plot/plot_2d.py

```diff
@@ -282,20 +294,63 @@ def update_slice(self, change):
                     self.slider_x[self.name][dim], val.value)
-                vslice = vslice[val.dim, val.value]
+                dslice = dslice[val.dim, val.value]
                 # At this point, after masks were combined, all their
                 # dimensions should be contained in the data_array.dims.
                 if self.params["masks"][self.name]["show"]:
                     mslice = mslice[val.dim, val.value]
             else:
-                button_dims[self.buttons[dim].value.lower() == "x"] = val.dim
+                # Get the dimensions of the dimension-coordinates, since
+                # buttons can contain non-dimension coordinates
+                button_dims[self.buttons[dim].value.lower() ==
+                            "x"] = self.slider_x[self.name][val.dim].dims[0]
 
         # Check if dimensions of arrays agree, if not, plot the transpose
-        slice_dims = vslice.dims
-        transp = slice_dims != button_dims
+        transp = dslice.dims != button_dims
+
+        # In the case of unaligned data, we may want to auto-scale the colorbar
+        # as we slice through dimensions. Colorbar limits are allowed to grow
+        # but not shrink.
+        autoscale_cbar = False
+        if dslice.unaligned is not None:
+            dslice = sc.histogram(dslice)
+            autoscale_cbar = True
+
+        # Make a new slice with bin edges and counts (for rebin), and with
+        # non-dimension coordinates if requested.
+        xy = "xy"
+        vslice = sc.DataArray(coords={
+            self.xyedges["x"].dims[0]: self.xyedges["x"],
+            self.xyedges["y"].dims[0]: self.xyedges["y"]
+        },
+                              data=sc.Variable(dims=[
+                                  self.xyedges[xy[not transp]].dims[0],
+                                  self.xyedges[xy[transp]].dims[0]
+                              ],
+                                               values=dslice.values,
+                                               variances=dslice.variances,
+                                               unit=sc.units.counts))
+
+        # Also include the masks
+        if self.params["masks"][self.name]["show"]:
+            mslice_dims = []
+            for dim in mslice.dims:
+                if dim == button_dims[0]:
+                    mslice_dims.append(self.xyedges["y"].dims[0])
+                elif dim == button_dims[1]:
+                    mslice_dims.append(self.xyedges["x"].dims[0])
+                else:
+                    mslice_dims.append(dim)
+            vslice.masks["all"] = sc.Variable(dims=mslice_dims,
+                                              values=mslice.values)
+
+        # Scale by bin width and then rebin in both directions
+        vslice *= self.xywidth["x"] * self.xywidth["y"]
+        vslice = sc.rebin(vslice, self.xyrebin["x"].dims[0], self.xyrebin["x"])
+        vslice = sc.rebin(vslice, self.xyrebin["y"].dims[0], self.xyrebin["y"])
```

Comment:
So are we *always* doing the rebin step, even if we have equal bins?

Reply:
yes, i will change to only rebinning when necessary